src/libcore/cell.rs

```diff
@@ -498,6 +460,90 @@ impl<T> Cell<T> {
     }
 }
 
+impl<T: ?Sized> Cell<T> {
+    /// Returns a raw pointer to the underlying data in this cell.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::cell::Cell;
+    ///
+    /// let c = Cell::new(5);
+    ///
+    /// let ptr = c.as_ptr();
+    /// ```
+    #[inline]
+    #[stable(feature = "cell_as_ptr", since = "1.12.0")]
+    pub fn as_ptr(&self) -> *mut T {
+        self.value.get()
+    }
+
+    /// Returns a mutable reference to the underlying data.
+    ///
+    /// This call borrows `Cell` mutably (at compile-time) which guarantees
+    /// that we possess the only reference.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::cell::Cell;
+    ///
+    /// let mut c = Cell::new(5);
+    /// *c.get_mut() += 1;
+    ///
+    /// assert_eq!(c.get(), 6);
+    /// ```
+    #[inline]
+    #[stable(feature = "cell_get_mut", since = "1.11.0")]
+    pub fn get_mut(&mut self) -> &mut T {
+        unsafe {
+            &mut *self.value.get()
+        }
+    }
+
+    /// Returns a `&Cell<T>` from a `&mut T`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(as_cell)]
+    /// use std::cell::Cell;
+    /// let slice: &mut [i32] = &mut [1,2,3];
+    /// let cell_slice: &Cell<[i32]> = Cell::from_mut(slice);
+    ///
+    /// assert_eq!(cell_slice.get_with(|v|v.len()), 3)
+    /// ```
+    #[inline]
+    #[unstable(feature = "as_cell", issue="43038")]
+    pub fn from_mut<'a>(t: &'a mut T) -> &'a Cell<T> {
+        unsafe {
+            &*(t as *mut T as *const Cell<T>)
+        }
+    }
+
+    /// Returns a value by applying a function on contained value
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(as_cell)]
+    /// use std::cell::Cell;
+    /// let c : Cell<Vec<i32>> = Cell::new(vec![1,2,3]);
+    /// let sum : i32 = c.get_with(|v|v.iter().sum());
+    /// assert_eq!(sum, 6_i32);
+    /// ```
+    #[inline]
+    #[unstable(feature = "as_cell", issue="43038")]
+    pub fn get_with<U, F>(&self, f: F) -> U
+    where
+        F: Fn(&T) -> U, U: 'static
```

Comment:
* `FnOnce` should be used in a situation like this.

* `U: 'static` doesn't do anything useful here, `U` can't hold the reference to `T` anyway in Rust, because `F` gets no guarantee as to how long that anonymous lifetime is.

* I'll make a separate comment on the PR about the unsoundness so it doesn't get lost